src/main/java/Matrices/ClusterAlgorithms/ART1.java

```diff
@@ -0,0 +1,239 @@
+
+    ArrayList<Integer> itemMemberships;
+    ArrayList<ArrayList<Double>> prototypes;
+    ArrayList<Grouping> prototypeGroups;
+
+
+    public ART1(SymmetricDSMData inputMatrix) {
+        matrix = inputMatrix.createCopy();
+        sortedItems = new ArrayList<>(matrix.getRows());
+
+        numItems = matrix.getRows().size();
+        adjacencyMatrix = new ArrayList<>();
+        for(int row = 0; row < numItems; row++) {
+            ArrayList<Double> matrixRow = new ArrayList<>();
+            for(int col = 0; col < numItems; col++) {
+                // if connection exists between this row and this column add a 1 to the matrix, otherwise, add a 0
+                if (matrix.getConnection(sortedItems.get(row).getUid(), sortedItems.get(col).getAliasUid()) != null) {
+                    matrixRow.add(1.0);
+                } else {
+                    matrixRow.add(0.0);
+                }
+            }
+            adjacencyMatrix.add(matrixRow);
+        }
+
+        // set all items to not having a group
+        matrix.clearGroupings();
+        itemMemberships = new ArrayList<>();
+        prototypes = new ArrayList<>();  // list of column uids where the prototype has a value
+        prototypeGroups = new ArrayList<>();
+    }
+
+
+    private void initPrototypes(int maxGroups) {
+        // create groupings with distinct colors
+        RandomColorGenerator rgc = new RandomColorGenerator(0.2423353);
+        for(int i = 0; i < maxGroups; i++) {
+            Grouping group = new Grouping("G" + i, null);
+
+            group.setColor(rgc.next());
+            prototypeGroups.add(group);  // add the group
+
+            prototypes.add(new ArrayList<>());  // add an empty list for the prototypes
+        }
+
+        for(int i = 0; i < numItems; i++) {
+            itemMemberships.add(-1);
+        }
+    }
+
+
+    /**
+     * performs an index-by-index fuzzy and of two vectors. ex <1, 0, 1> and <1, 0, 0> = <1, 0, 0>
+     * @param v1 the first vector
+     * @param v2 the second vector
+     * @return the resulting vector
+     */
+    private ArrayList<Double> andVectors(ArrayList<Double> v1, ArrayList<Double> v2) {
+        ArrayList<Double> v = new ArrayList<>();
+        for(int i = 0; i < v1.size(); i++) {
+            if(v1.get(i) < v2.get(i)) {  // fuzzy and operator -- choose minimum
+                v.add(v1.get(i));
+            } else {
+                v.add(v2.get(i));
+            }
+        }
+
+        return v;
+    }
+
+
+    private ArrayList<Double> sumVectors(ArrayList<Double> v1, ArrayList<Double> v2) {
+        ArrayList<Double> v = new ArrayList<>();
+        for(int i = 0; i < v1.size(); i++) {
+            v.add(v1.get(i) + v2.get(i));
+        }
+
+        return v;
+    }
+
+
+    /**
+     * returns magnitude of vector defined as sum of all elements
+     * @param v vector
+     * @return magnitude
+     */
+    private double vectorMagnitude(ArrayList<Double> v) {
+        int sum = 0;
```

Comment:
Here may occur int overhead, may be need use other structure, BigInteger for example